overlord/exportstate/manifest.go

```diff
@@ -0,0 +1,211 @@
+	var firstErr error
+	for _, s := range m.Symlinks {
+		if err := s.Remove(); err != nil && firstErr == nil {
+			firstErr = err
+		}
+	}
+	return firstErr
+}
+
+// SymlinkExport describes content exported as symbolic link.
+type SymlinkExport struct {
+	SnapName        string `json:"snap-name"`
+	ExportedVersion string `json:"exported-version"`
+	ExportSet       string `json:"export-set"`
+	Name            string `json:"name"`
+	Target          string `json:"target"`
+}
+
+// PathName returns the full path of the symbolic link.
+func (s *SymlinkExport) PathName() string {
+	return filepath.Join(ExportDir, s.SnapName, s.ExportedVersion, s.ExportSet, s.Name)
+}
+
+// Create creates a symbolic link and necessary directories.
+//
+// The function is idempotent.
+func (s *SymlinkExport) Create() error {
+	pathName := s.PathName()
+	if err := os.MkdirAll(filepath.Dir(pathName), 0755); err != nil {
+		return err
+	}
+	err := os.Symlink(s.Target, pathName)
+	if err != nil && os.IsExist(err) {
+		if actualTarget, _ := os.Readlink(pathName); actualTarget == s.Target {
+			err = nil
+		}
+	}
+	return err
+}
+
+// Remove removes the symbolic link and prunes any directories.
+func (s *SymlinkExport) Remove() error {
+	if err := os.Remove(s.PathName()); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// XXX: or iterate upwards until we reach ExportDir
+	os.Remove(filepath.Join(ExportDir, s.SnapName, s.ExportedVersion, s.ExportSet))
+	os.Remove(filepath.Join(ExportDir, s.SnapName, s.ExportedVersion))
+	os.Remove(filepath.Join(ExportDir, s.SnapName))
+	return nil
+}
+
+// exportSetSymlinks returns symlink exports for given snap and export entries.
+func exportSetSymlinks(snapName, snapRev, exportSetName string, entries []*exportEntry) []SymlinkExport {
+	symlinks := make([]SymlinkExport, 0, len(entries))
+	for _, entry := range entries {
+		var target string
+		if entry.IsExportedPathValidInHostMountNS {
+			target = entry.PathInHostMountNS
+		} else {
+			target = entry.PathInSnapMountNS
+		}
+		symlinks = append(symlinks, SymlinkExport{
+			SnapName:        snapName,
+			ExportedVersion: snapRev,
+			ExportSet:       exportSetName,
+			Name:            entry.PathInExportSet,
+			Target:          target,
+		})
+	}
+	return symlinks
+}
+
+// exportEntry describes a single file placed in a specific export set.
+//
+// The original file is described by two paths. PathInHostMountNS is is valid
+// in the host mount namespace while PathInSnapMountNS is valid in the per-snap
+// mount namespace. The original file is exported by creating a symbolic link
+// under PathInExportSet, which is relative to the export set that the entry
+// belongs to, to one of the two other paths.
+//
+// If IsExportedPathValidInHostMountNS returns true then PathInHostMountNS is
+// used, if it returns false PathInSnapMountNS is used instead.
+//
+// This distinction enables exporting files that are consumed by either other
+// snaps or by the classic system.
```

Comment:
Indeed but that was my goal. For as long as export sets are the thing we can switch, I did not want to limit their expressiveness. At some point, snapd snap may export files for the "host", for example if we want to solve the "snap command is broken while snapd is unlinked". Right how /usr/bin/snap requires a linked snapd snap, we could instead point it at an export set, allowing it to be linked at all times.

Reply:
the goal is fine, but now that I understand a bit more the intention/code, I think the implementation is a bit too opaque and flexible,  we need to reorg differently Manifest.Symlinks so that the source/target context are clearer, and also we don't need to repeat info or store it too eagerly.

Reply:
Let's discuss the details of this at the call in 30 minutes and I'll make the changes today.

Reply:
We are now in agreement as to how this should look like.